working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,166 @@
+
+In NNBD libraries, we do not require that *final* instance variables are always initialized, but we do require that *potentially non-nullable* variables are. We could require that final fields are initialized, but it's very reasonable to treat being *nullable* as meaning being optional. Given that there is an initializing formal parameter for the field, it will definitely be initialized
+
+An instance variable in a subclass will always shadow a named parameter of the same name in a superclass constructor. That's why we do not forward to superclass constructors which have a required named parameter with such a name.
+
+Superclass constructors with optional positional parameters are hard to forward to. It would be wonderful if it was possible to have both optional positional parameters and named parameters on the same function.
+
+Existing mixin-application class forwarding constructors are just a special case of the general constructor forwarding functionality.
+
+We make constructors `const` only when the subclass declares no instance variables. We could allow final instance variables with no initializer expression (because then the constructor must initialize it) or with an initializer expression which is constant (because then it's valid in a `const` constructor class). The latter requirement is particularly fragile and useless. Changing the initializer expression of a variable from `final int x = a + 2;` to `final int x = a + b;` where `a` is a constant variable and  `b` is a non-constant variable with the value `2` will change the constructors of the class from constant to non-constant. Also, a final instance variable initialized to a constant value should probably just be a getter. We want to avoid accidentally making a class have constant constructors, because that may become a support burden for the author, so we only make constructors constant when the subclass is particularly simple, and where the current mixin application constructor forwarding makes the constructor constant. (It is not a *breaking* change to make more constructors constant in the future, but it is a *bad* change for library owners which intended a class to have non-constant constructors).
+
+You only get default constructors when you write *no* constructors. If you want to forward ten superclass constructors, and change the eleventh just a bit, or add an eleventh factory constructor, then you have to write all of the other ten in full. If this is too annoying, we may be able to add brief syntax for forwarding a constructor verbatim, say `Foo.bar = super.bar`.
```

Comment:
+1 to it being annoying. When possible, I like to avoid features where one minor change necessitates a cascade of other changes.

Reply:
OK, I'm convinced. Will rewrite to do that.